linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+         }
+      }
+      return dual<value_type, gradient_type> {value, gradient};
+   });
+}
+
+/**
+ * @brief recursively serialize the entries in a tensor to an output stream.
+ * Output format uses braces and comma separators to mimic C syntax for multidimensional array
+ * initialization.
+ *
+ * @param[in] os The stream to work with standard output streams
+ * @param[in] A The tensor to write out
+ */
+template <typename T, int... n>
+std::ostream& operator<<(std::ostream& os, const tensor<T, n...>& A)
+{
+   os << '{' << A[0];
+   for (int i = 1; i < tensor<T, n...>::first_dim; i++)
+   {
+      os << ", " << A[i];
+   }
+   os << '}';
+   return os;
+}
+
+/**
+ * @brief replace all entries in a tensor satisfying |x| < 1.0e-10 by literal zero
+ * @param[in] A The tensor to "chop"
+ */
+template <int n> MFEM_HOST_DEVICE
+tensor<double, n> chop(const tensor<double, n>& A)
+{
+   auto copy = A;
+   for (int i = 0; i < n; i++)
+   {
+      if (copy[i] * copy[i] < 1.0e-20)
+      {
+         copy[i] = 0.0;
+      }
+   }
+   return copy;
+}
+
+/// @overload
+template <int m, int n> MFEM_HOST_DEVICE
+tensor<double, m, n> chop(const tensor<double, m, n>& A)
+{
+   auto copy = A;
+   for (int i = 0; i < m; i++)
+   {
+      for (int j = 0; j < n; j++)
+      {
+         if (copy[i][j] * copy[i][j] < 1.0e-20)
+         {
+            copy[i][j] = 0.0;
+         }
+      }
+   }
+   return copy;
+}
+
+/// @cond
+namespace detail
+{
+template <typename T1, typename T2>
+struct outer_prod;
+
+template <int... m, int... n>
+struct outer_prod<tensor<double, m...>, tensor<double, n...>>
+{
+   using type = tensor<double, m..., n...>;
+};
+
+template <int... n>
+struct outer_prod<double, tensor<double, n...>>
+{
+   using type = tensor<double, n...>;
+};
+
+template <int... n>
+struct outer_prod<tensor<double, n...>, double>
+{
+   using type = tensor<double, n...>;
+};
+
+template <>
+struct outer_prod<double, double>
+{
+   using type = tensor<double>;
+};
+
+template <typename T>
+struct outer_prod<zero, T>
+{
+   using type = zero;
+};
+
+template <typename T>
+struct outer_prod<T, zero>
+{
+   using type = zero;
+};
+
+}  // namespace detail
+/// @endcond
+
+/**
+ * @brief a type function that returns the tensor type of an outer product of two tensors
+ * @tparam T1 the first argument to the outer product
+ * @tparam T2 the second argument to the outer product
+ */
+template <typename T1, typename T2>
+using outer_product_t = typename detail::outer_prod<T1, T2>::type;
+
+/**
+ * @brief Retrieves the gradient component of a double (which is nothing)
+ * @return The sentinel, @see zero
+ */
+inline MFEM_HOST_DEVICE zero get_gradient(double /* arg */) { return zero{}; }
+
+/**
+ * @brief get the gradient of type `tensor` (note: since its stored type is not a dual
+ * number, the derivative term is identically zero)
+ * @return The sentinel, @see zero
+ */
+template <int... n>
+MFEM_HOST_DEVICE zero get_gradient(const tensor<double, n...>& /* arg */)
```

Comment:
Do we need these `get_gradient` functions?

Reply:
yes, they enable the `tensor` and `dual` classes to be composed